README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

A Flutter library to calculate Internal Rate Of Return(XIRR) for financial investments.

## Features

Calculates Internal Rate Of Return(XIRR) to know return on investment. It uses Newton-Raphson method
for calculating Internal Rate Of Return

## Getting started

Library will throw exceptions if inputs are not in line with expectations

* If all transactions are positive/negative only
* If all transactions are for same day
* If no transactions are passed

Please check test cases to know more details

## Usage

Deposit/Investment should be provided with negative amount and withdrawals should be provided with
positive amount.

```dart

List<Transaction> transactions = [];

transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
transactions.add(Transaction.withStringDate(1100, "2011-01-01"));

final double? result =
XirrFlutter.withTransactions(transactions).calculate();
expect(result, 0.10);

```

## Additional information

This library was created based on Java Xirr library available
at https://github.com/RayDeCampo/java-xirr
Hence terms of the library and license follows original creator

CHANGELOG.md
## 1.0.0

* Initial version of the library.

## 1.0.1

* Minor changes to the project and few bug fixes


## 1.0.2

* Minor changes to the project and few bug fixes

## 1.0.3

* Minor changes to the project and few bug fixes

test/xirr_flutter_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:xirr_flutter/src/exceptions.dart';
import 'package:xirr_flutter/xirr_flutter.dart';

void main() {
  test('One year no growth', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(1000, "2011-01-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result, 0);
  });

  test('One year growth', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(1100, "2011-01-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result, 0.10);
  });

  test('One year decline', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(900, "2011-01-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result, -0.10);
  });

  test('Compared to spreadsheet output', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-10-01"));
    transactions.add(Transaction.withStringDate(4300, "2011-01-01"));
    transactions.add(Transaction.withStringDate(-1000, "2010-07-01"));
    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(-1000, "2010-04-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result?.toPrecision(7), 0.1212676);
  });

  test('100 percent growth', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(3000, "2011-01-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result, 2.00);
  });

  test('Total loss in a year', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(0, "2011-01-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result, -1.00);
  });

  test('Total loss in 2 years', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(0, "2012-01-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result, -1.00);
  });

  test('Total loss in 1/2 year', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(0, "2010-07-01"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result, -1.00);
  });

  test('Calculation from Sample method 1', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2016-01-15"));
    transactions.add(Transaction.withStringDate(-2500, "2016-02-08"));
    transactions.add(Transaction.withStringDate(-1000, "2016-04-17"));
    transactions.add(Transaction.withStringDate(5050, "2016-08-24"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();
    expect(result?.toPrecision(11), 0.25042347105);
  });

  test('issue from node js version', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-10000, "2000-05-24"));
    transactions.add(Transaction.withStringDate(3027.25, "2000-06-05"));
    transactions.add(Transaction.withStringDate(630.68, "2001-04-09"));
    transactions.add(Transaction.withStringDate(2018.2, "2004-02-24"));
    transactions.add(Transaction.withStringDate(1513.62, "2005-03-18"));
    transactions.add(Transaction.withStringDate(1765.89, "2006-02-15"));
    transactions.add(Transaction.withStringDate(4036.33, "2007-01-10"));
    transactions.add(Transaction.withStringDate(4036.33, "2007-11-14"));
    transactions.add(Transaction.withStringDate(1513.62, "2008-12-17"));
    transactions.add(Transaction.withStringDate(1513.62, "2010-01-15"));
    transactions.add(Transaction.withStringDate(2018.16, "2011-01-14"));
    transactions.add(Transaction.withStringDate(1513.62, "2012-02-03"));
    transactions.add(Transaction.withStringDate(1009.08, "2013-01-18"));
    transactions.add(Transaction.withStringDate(1513.62, "2014-01-24"));
    transactions.add(Transaction.withStringDate(1513.62, "2015-01-30"));
    transactions.add(Transaction.withStringDate(1765.89, "2016-01-22"));
    transactions.add(Transaction.withStringDate(1765.89, "2017-01-20"));
    transactions.add(Transaction.withStringDate(22421.55, "2017-06-05"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();

    expect(result?.toPrecision(7), 0.2126861);
  });

  test('issue 5a', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-2610, "2001-06-22"));
    transactions.add(Transaction.withStringDate(-2589, "2001-07-03"));
    transactions.add(Transaction.withStringDate(-5110, "2001-07-05"));
    transactions.add(Transaction.withStringDate(-2550, "2001-07-06"));
    transactions.add(Transaction.withStringDate(-5086, "2001-07-09"));
    transactions.add(Transaction.withStringDate(-2561, "2001-07-10"));
    transactions.add(Transaction.withStringDate(-5040, "2001-07-12"));
    transactions.add(Transaction.withStringDate(-2552, "2001-07-13"));
    transactions.add(Transaction.withStringDate(-2530, "2001-07-16"));
    transactions.add(Transaction.withStringDate(29520, "2001-07-17"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();

    expect(result?.toPrecision(7), -0.7640294);
  });

  test('issue 5b', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-2610, "2001-06-22"));
    transactions.add(Transaction.withStringDate(-2589, "2001-07-03"));
    transactions.add(Transaction.withStringDate(-5110, "2001-07-05"));
    transactions.add(Transaction.withStringDate(-2550, "2001-07-06"));
    transactions.add(Transaction.withStringDate(-5086, "2001-07-09"));
    transactions.add(Transaction.withStringDate(-2561, "2001-07-10"));
    transactions.add(Transaction.withStringDate(-5040, "2001-07-12"));
    transactions.add(Transaction.withStringDate(-2552, "2001-07-13"));
    transactions.add(Transaction.withStringDate(-2530, "2001-07-16"));
    transactions.add(Transaction.withStringDate(-9840, "2001-07-17"));
    transactions.add(Transaction.withStringDate(38900, "2001-07-18"));

    final double? result =
        XirrFlutter.withTransactions(transactions).calculate();

    expect(result, -0.8353404468272648);
  });

  test('Transaction with guess parameter', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-2610, "2001-06-22"));
    transactions.add(Transaction.withStringDate(-2589, "2001-07-03"));
    transactions.add(Transaction.withStringDate(-5110, "2001-07-05"));
    transactions.add(Transaction.withStringDate(-2550, "2001-07-06"));
    transactions.add(Transaction.withStringDate(-5086, "2001-07-09"));
    transactions.add(Transaction.withStringDate(-2561, "2001-07-10"));
    transactions.add(Transaction.withStringDate(-5040, "2001-07-12"));
    transactions.add(Transaction.withStringDate(-2552, "2001-07-13"));
    transactions.add(Transaction.withStringDate(-2530, "2001-07-16"));
    transactions.add(Transaction.withStringDate(-9840, "2001-07-17"));
    transactions.add(Transaction.withStringDate(38900, "2001-07-18"));

    final double? result =
        XirrFlutter.withTransactionsAndGuess(transactions, -0.90).calculate();

    expect(result, -0.835340446830991);
  });

  test('issue 23 is impossible', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-2000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(2508.80, "2012-01-01"));

    double? result = XirrFlutter.withTransactions(transactions).calculate();

    expect(result?.toPrecision(7), 0.12);

    transactions.clear();
    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(-1120, "2011-01-01"));
    transactions.add(Transaction.withStringDate(2508.8, "2012-01-01"));

    result = XirrFlutter.withTransactions(transactions).calculate();

    expect(result?.toPrecision(7), 0.12);

    transactions.clear();
    transactions.add(Transaction.withStringDate(-2000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(2508.80, "2013-01-01"));

    result = XirrFlutter.withTransactions(transactions).calculate();

    expect(result?.toPrecision(7), 0.0784055);

    transactions.clear();
    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(-1120, "2011-01-01"));
    transactions.add(Transaction.withStringDate(2508.8, "2013-01-01"));

    result = XirrFlutter.withTransactions(transactions).calculate();

    expect(result?.toPrecision(8), 0.07017196);
  });

  test('Negative: No transactions exception', () {
    List<Transaction> transactions = [];

    expect(() => XirrFlutter.withTransactions(transactions),
        throwsA(isA<MissingData>()));
  });

  test('Negative: Single transaction only exception', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));

    expect(() => XirrFlutter.withTransactions(transactions),
        throwsA(isA<MissingData>()));
  });

  test('Negative: All transactions for same day exception', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01 09:00"));
    transactions.add(Transaction.withStringDate(-1000, "2010-01-01 12:00"));
    transactions.add(Transaction.withStringDate(2100, "2010-01-01 15:00"));

    expect(() => XirrFlutter.withTransactions(transactions).calculate(),
        throwsA(isA<InvalidDataForSameDay>()));
  });

  test('Negative: All transactions are negative exception', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(-1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(-1000, "2010-05-01"));
    transactions.add(Transaction.withStringDate(-2000, "2010-09-01"));

    expect(() => XirrFlutter.withTransactions(transactions),
        throwsA(isA<AllNegativeAmounts>()));
  });

  test('Negative: All transactions are with positive amount exception', () {
    List<Transaction> transactions = [];

    transactions.add(Transaction.withStringDate(1000, "2010-01-01"));
    transactions.add(Transaction.withStringDate(1000, "2010-05-01"));
    transactions.add(Transaction.withStringDate(0, "2010-09-01"));

    expect(() => XirrFlutter.withTransactions(transactions),
        throwsA(isA<AllNonNegativeAmounts>()));
  });
}


test/newton_raphson_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:xirr_flutter/src/exceptions.dart';
import 'package:xirr_flutter/src/newton_raphson.dart';
import 'package:xirr_flutter/xirr_flutter.dart';

void main() {
  test('Square root', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return value * value;
    }, (final double value) {
      return 2 * value;
    });

    expect(newtonRaphson.inverse(4, 4)?.toPrecision(0), 2);
    expect(newtonRaphson.inverse(9, -9)?.toPrecision(0), -3);
    expect(newtonRaphson.inverse(625, 625)?.toPrecision(0), 25);
  });

  test('Cube root', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return value * value * value;
    }, (final double value) {
      return 3 * value * value;
    });

    expect(newtonRaphson.inverse(8, 8)?.toPrecision(0), 2);
    expect(newtonRaphson.inverse(-27, 27)?.toPrecision(0), -3);
    expect(newtonRaphson.inverse(15625, 15625)?.toPrecision(0), 25);
  });

  test('Quadratic root', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return (value - 4) * (value + 3);
    }, (final double value) {
      return 2 * value - 1;
    });

    expect(newtonRaphson.findRoot(10)?.toPrecision(0), 4);
    expect(newtonRaphson.findRoot(-10)?.toPrecision(0), -3);
    expect(newtonRaphson.findRoot(0.51)?.toPrecision(0), 4);
    expect(newtonRaphson.findRoot(0.49)?.toPrecision(0), -3);
  });

  test('Fail to coverage', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return (value - 4) * (value + 3);
    }, (final double value) {
      return 2 * value - 1;
    });

    expect(() => newtonRaphson.findRoot(0.5),
        throwsA(isA<ZeroValuedDerivativeException>()));
  });

  test('Fail to coverage with Verification', () {
    NewtonRaphson nr = NewtonRaphson.withFunctions((final double value) {
      return 2;
    }, (final double value) {
      return value > 0 ? 0.25 : 0;
    });

    try {
      nr.findRoot(3);
    } on ZeroValuedDerivativeException catch (e) {
      expect(e.getInitialGuess(), 3);
      expect(e.getIteration(), 2);
      expect(e.getCandidate(), -5);
      expect(e.getValue(), 2);
    }
  });

  test('Fail to coverage non convergence', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return 2 * value.sign;
    }, (final double value) {
      return 1;
    });

    expect(() => newtonRaphson.findRoot(1),
        throwsA(isA<NonConvergenceException>()));
  });

  test('Fail to coverage with bad candidate', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return double.maxFinite;
    }, (final double value) {
      return double.minPositive;
    });

    try {
      newtonRaphson.findRoot(3);
    } on CandidateIsInfinite catch (e) {
      expect(e.getInitialGuess(), 3);
      expect(e.getIteration(), 1);
      expect(double.negativeInfinity, e.getCandidate());
    }
  });

  test('Fail to coverage NaN function value', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return double.nan;
    }, (final double value) {
      return 1;
    });

    try {
      newtonRaphson.findRoot(3);
    } on FunctionValueOverflow catch (e) {
      expect(e.getInitialGuess(), 3);
      expect(e.getIteration(), 1);
      expect(e.getCandidate(), 3);
      expect(e.getValue()?.isNaN, true);
      expect(e.getDerivedValue(), isNull);
    }
  });

  test('Fail to coverage NaN derivative value', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctions((final double value) {
      return 2;
    }, (final double value) {
      return double.nan;
    });

    try {
      newtonRaphson.findRoot(3);
    } on DerivativeValueOverflow catch (e) {
      expect(e.getInitialGuess(), 3);
      expect(e.getIteration(), 1);
      expect(e.getCandidate(), 3);
      expect(e.getValue(), 2);
      expect(e.getDerivedValue()?.isNaN, true);
    }
  });

  test('Tolerance', () {
    NewtonRaphson newtonRaphson =
        NewtonRaphson.withFunctionsAndTolerance((final double value) {
      return value * value;
    }, (final double value) {
      return 2 * value;
    }, NewtonRaphson.defaultTolerance / 1000);

    expect(newtonRaphson.inverse(16, 16)?.toPrecision(0), 4);
    expect(newtonRaphson.inverse(225, 225)?.toPrecision(0), 15);
    expect(newtonRaphson.inverse(2, 2)?.toPrecision(10), 1.4142135624);
  });
}


lib/transaction.dart
library xirr_flutter;

import 'dart:core';

/// Represents a transaction for the purposes of computing the irregular rate
/// of return.
/// <p>
/// Note that negative amounts represent deposits into the investment (and so
/// withdrawals from your cash).  Positive amounts represent withdrawals from the
/// investment (deposits into cash).  Zero amounts are allowed in case your
/// investment is now worthless.
/// @see Xirr
class Transaction {
  late final double amount;
  late final DateTime when;

  /// Construct a Transaction instance with the given amount at the given day.
  /// @param amount the amount transferred
  /// @param when the day the transaction took place
  Transaction(this.amount, this.when);

  /// Construct a Transaction instance with the given amount at the given day.
  /// @param amount the amount transferred
  /// @param when the day the transaction took place
  Transaction.withStringDate(this.amount, String when) {
    DateTime temp = DateTime.parse(when);
    this.when = DateTime(temp.year, temp.month, temp.day);
  }
}


lib/src/exceptions.dart
import 'package:xirr_flutter/src/calculation.dart';

class OverflowException implements Exception {
  late final Calculation _state;

  OverflowException(this._state);

  double? getInitialGuess() {
    return _state.getGuess();
  }

  int getIteration() {
    return _state.getIteration();
  }

  double? getCandidate() {
    return _state.candidate;
  }

  double? getValue() {
    return _state.value;
  }

  double? getDerivedValue() {
    return _state.derivativeValue;
  }

  @override
  toString() {
    return "${super.toString()}: $_state";
  }
}

class NonConvergenceException implements Exception {
  final double _initialGuess;
  final int _iterations;

  NonConvergenceException(this._initialGuess, this._iterations);

  double getInitialGuess() {
    return _initialGuess;
  }

  int getIterations() {
    return -_iterations;
  }

  String errorMessage() {
    return "Newton-Raphson failed to converge within $_iterations iterations.";
  }
}

class MissingData implements Exception {
  String errorMessage() {
    return "Must have at least two transactions";
  }
}

class InvalidDataForSameDay implements Exception {
  String errorMessage() {
    return "Transactions must not all be on the same day.";
  }
}

class CandidateIsInfinite extends OverflowException {
  CandidateIsInfinite(super.state);

  String errorMessage() {
    return "Candidate is infinite";
  }
}

class FunctionValueOverflow extends OverflowException {
  FunctionValueOverflow(super.state);

  String errorMessage() {
    return "Function value overflow";
  }
}

class DerivativeValueOverflow extends OverflowException {
  DerivativeValueOverflow(super.state);

  String errorMessage() {
    return "Derivative value overflow";
  }
}

class AllNegativeAmounts implements Exception {
  String errorMessage() {
    return "Transactions must not be negative.";
  }
}

class AllNonNegativeAmounts implements Exception {
  String errorMessage() {
    return "Transactions must not all be non-negative.";
  }
}

class ZeroValuedDerivativeException extends OverflowException {
  ZeroValuedDerivativeException(super.state);

  String errorMessage() {
    return "Newton-Raphson failed due to zero-valued derivative.";
  }
}


lib/src/investment.dart
import 'dart:core';
import 'dart:math';

/// Convenience class which represents {@link Transaction} instances more
/// conveniently for our purposes.
class Investment {
  /// The amount of the investment.
  late double amount;

  /// The number of years for which the investment applies, including
  /// fractional years.
  late double years;

  /// Present value of the investment at the given rate.
  /// @param rate the rate of return
  /// @return present value of the investment at the given rate
  double presentValue(final double rate) {
    if (-1 < rate) {
      return amount * pow(1 + rate, years);
    } else if (rate < -1) {
// Extend the function into the range where the rate is less
// than -100%.  Even though this does not make practical sense,
// it allows the algorithm to converge in the cases where the
// candidate values enter this range

// We cannot use the same formula as before, since the base of
// the exponent (1+rate) is negative, this yields imaginary
// values for fractional years.
// E.g. if rate=-1.5 and years=.5, it would be (-.5)^.5,
// i.e. the square root of negative one half.

// Ensure the values are always negative so there can never
// be a zero (as long as some amount is non-zero).
// This formula also ensures that the derivative is positive
// (when rate < -1) so that Newton's method is encouraged to
// move the candidate values towards the proper range

      return -(amount.abs()) * pow(-1 - rate, years);
    } else if (years == 0) {
      return amount; // Resolve 0^0 as 0
    } else {
      return 0;
    }
  }

  /// Derivative of the present value of the investment at the given rate.
  /// @param rate the rate of return
  /// @return derivative of the present value at the given rate
  double derivative(final double rate) {
    if (years == 0) {
      return 0;
    } else if (-1 < rate) {
      return amount * years * pow(1 + rate, years - 1);
    } else if (rate < -1) {
      return (amount.abs()) * years * pow(-1 - rate, years - 1);
    } else {
      return 0;
    }
  }
}


lib/src/calculation.dart
import 'dart:core';

import 'exceptions.dart';

class Calculation {
  double? guess;
  int iteration = 0; // persistent loop counter
  double? candidate;
  double? value;
  double? derivativeValue;
  late final double Function(double) func;
  late final double Function(double) derivative;
  late final double tolerance;
  late final int iterations;

  Calculation(this.func, this.derivative, this.iterations, this.tolerance);

  double? getGuess() {
    return guess;
  }

  void setGuess(double guess) {
    this.guess = guess;
  }

  int getIteration() {
    return iteration + 1;
  }

  double? getCandidate() {
    return candidate;
  }

  void setCandidate(double candidate) {
    this.candidate = candidate;
    if (!(candidate.isFinite)) {
      throw CandidateIsInfinite(this);
    }
  }

  double? getValue() {
    return value;
  }

  void setValue(double value) {
    this.value = value;
    if (!(value.isFinite)) {
      throw FunctionValueOverflow(this);
    }
  }

  double? getDerivativeValue() {
    return derivativeValue;
  }

  void setDerivativeValue(double derivativeValue) {
    this.derivativeValue = derivativeValue;
    if (!(derivativeValue.isFinite)) {
      throw DerivativeValueOverflow(this);
    } else if (derivativeValue == 0.0) {
      throw ZeroValuedDerivativeException(this);
    }
  }

  @override
  String toString() {
    return "{ guess=$guess, iteration=$iteration, candidate=$candidate, value=$value, derivative=$derivativeValue }";
  }

  double? solve(double guess, double target) {
    setGuess(guess);
    setCandidate(guess);
    for (iteration = 0; iteration < iterations; iteration++) {
      setValue(func.call(candidate!) - target);
      if ((value?.abs())! < tolerance) {
        return candidate;
      } else {
        setDerivativeValue(derivative.call(candidate!));
        setCandidate(candidate! - value! / derivativeValue!);
      }
    }
    throw NonConvergenceException(guess, iterations);
  }
}


lib/src/xirr_details.dart
import 'dart:math';

import '../transaction.dart';
import 'exceptions.dart';

/// Converts a stream of {@link Transaction} instances into the data needed for
/// the {@link Xirr_Flutter} algorithm.
class XirrDetails {
  DateTime? start;
  DateTime? end;
  double minAmount = double.infinity;
  double maxAmount = double.negativeInfinity;
  double? total;
  double? deposits;

  void accumulate(final Transaction tx) {
    start =
        start != null ? (start!.isBefore(tx.when) ? start : tx.when) : tx.when;
    end = end != null ? (end!.isAfter(tx.when) ? end : tx.when) : tx.when;
    minAmount = min(minAmount, tx.amount);
    maxAmount = max(maxAmount, tx.amount);
    total = total == null ? tx.amount : total! + tx.amount;
    if (tx.amount < 0) {
      deposits = deposits == null ? tx.amount : deposits! - tx.amount;
    }
  }

  void validate() {
    if (start == null) {
      throw Exception("No transactions to analyze");
    }

    if (start == end) {
      throw InvalidDataForSameDay();
    }
    if (minAmount >= 0) {
      throw AllNonNegativeAmounts();
    }
    if (maxAmount < 0) {
      throw AllNegativeAmounts();
    }
  }
}


lib/src/newton_raphson.dart
import 'dart:core';

import 'calculation.dart';

/// Simple implementation of the Newton-Raphson method for finding roots or
/// inverses of a function.
/// <p>
/// The function and its derivative must be supplied as instances of

/// Function and the answers are computed as doubles.
/// <p>
/// For examples of usage, see the source of the test class or the Xirr_Flutter class.
/// <p>
/// The <code>iterations</code> parameter is used as an upper bound on the number
/// of iterations to run the method for.
/// <p>
/// The <code>tolerance</code> parameter is used to determine when the method
/// has been successful.  If the value of the function at the candidate input
/// is within the <code>tolerance</code> of the desired target value, the
/// method terminates.
class NewtonRaphson {
  /// Default tolerance.
  static const double defaultTolerance = 0.0000001;
  static const int defaultMaxIterations = 10000;
  late final double Function(double) func;
  late final double Function(double) derivative;
  late final double tolerance;
  late final int iterations;

  /// Construct an instance of the NewtonRaphson method
  /// @param func the function
  /// @param derivative the derivative of the function
  /// @param tolerance the tolerance
  /// @param iterations maximum number of iterations
  NewtonRaphson(this.func, this.derivative, this.tolerance, this.iterations);

  NewtonRaphson.withFunctions(this.func, this.derivative) {
    tolerance = defaultTolerance;
    iterations = defaultMaxIterations;
  }

  NewtonRaphson.withFunctionsAndTolerance(
      this.func, this.derivative, this.tolerance) {
    iterations = defaultMaxIterations;
  }

  /// Equivalent to <code>inverse(0, guess)</code>.
  /// <p>
  /// Find a root of the function starting at the given .  Equivalent to
  /// invoking <code>inverse(0, guess)</code>.  Finds the input value <i>x</i>
  /// such that |<i>f</i>(<i>x</i>)| &lt; <i>tolerance</i>.
  /// @param guess the value to start at
  /// @return an input to the function which yields zero within the given
  ///         tolerance
  /// @see #inverse(double, double)
  double? findRoot(final double guess) {
    return inverse(0, guess);
  }

  /// Find the input value to the function which yields the given
  /// <code>target</code>, starting at the <code>guess</code>.  More precisely,
  /// finds an input value <i>x</i>
  /// such that |<i>f</i>(<i>x</i>) - <code>target</code>| &lt; <i>tolerance</i>
  /// @param target the target value of the function
  /// @param guess value to start the algorithm with
  /// @return the inverse of the function at <code>target</code> within the
  /// given tolerance
  /// @throws Exception if the derivative is 0 while executing the Newton-Raphson method
  /// @throws Exception when a value involved is infinite or NaN
  /// @throws Exception if the method fails to converge in the given number of iterations
  double? inverse(final double target, final double guess) {
    return Calculation(func, derivative, iterations, tolerance)
        .solve(guess, target);
  }
}


lib/xirr_flutter.dart
library xirr_flutter;

export 'transaction.dart';

import 'src/exceptions.dart';
import 'src/investment.dart';
import 'src/newton_raphson.dart';
import 'src/xirr_details.dart';
import 'transaction.dart';

/// Calculates the irregular rate of return on a series of transactions.  The
/// irregular rate of return is the constant rate for which, if the transactions
/// had been applied to an investment with that rate, the same resulting returns
/// would be realized.
/// <p>
/// When creating the list of {@link Transaction} instances to feed Xirr_Flutter, be
/// sure to include one transaction representing the present value of the account
/// now, as if you had cashed out the investment.
/// <p>
/// Example usage:
/// <code>
///     List<Transaction> transactions = [];
///
///     transactions.add(Transaction.withStringDate(-1000, "2016-01-15"));
///     transactions.add(Transaction.withStringDate(-2500, "2016-02-08"));
///     transactions.add(Transaction.withStringDate(-1000, "2016-04-17"));
///     transactions.add(Transaction.withStringDate(5050, "2016-08-24"));
///
///     final double? result =
///     XirrFlutter.withTransactions(transactions).calculate();
///     expect(result?.toPrecision(11), 0.25042347105);
/// </code>

class XirrFlutter {
  static const double daysInYear = 365;
  late final NewtonRaphson newtonRaphson;
  late final List<Investment> investments = [];
  late final XirrDetails details = XirrDetails();
  double? guess;

  XirrFlutter.withTransactions(List<Transaction> transactions) {
    processTransactions(transactions);
  }

  XirrFlutter.withTransactionsAndGuess(
      List<Transaction> transactions, this.guess) {
    processTransactions(transactions);
  }

  void processTransactions(List<Transaction> transactions) {
    if (transactions.length < 2) {
      throw MissingData();
    }
    for (var transaction in transactions) {
      details.accumulate(transaction);
    }
    details.validate();

    for (var transaction in transactions) {
      investments.add(createInvestment(transaction));
    }
  }

  static int daysBetween(DateTime from, DateTime to) {
    from = DateTime(from.year, from.month, from.day);
    to = DateTime(to.year, to.month, to.day);
    return (to.difference(from).inHours / 24).round();
  }

  Investment createInvestment(Transaction tx) {
    // Transform the transaction into an Investment instance
    // It is much easier to calculate the present value of an Investment
    final Investment result = Investment();
    result.amount = tx.amount;
    // Don't use YEARS.between() as it returns whole numbers
    result.years = daysBetween(tx.when, details.end!) / daysInYear;
    return result;
  }

  /// Calculates the present value of the investment if it had been subject to
  /// the given rate of return.
  /// @param rate the rate of return
  /// @return the present value of the investment if it had been subject to the
  ///         given rate of return
  double presentValue(final double rate) {
    return investments.fold(0,
        (previousValue, element) => element.presentValue(rate) + previousValue);
  }

  /// The derivative of the present value under the given rate.
  /// @param rate the rate of return
  /// @return derivative of the present value under the given rate
  double derivative(final double rate) {
    return investments.fold(0,
        (previousValue, element) => element.derivative(rate) + previousValue);
  }

  /// Calculates the irregular rate of return of the transactions for this
  /// instance of Xirr.
  /// @return the irregular rate of return of the transactions
  /// @throws Exception if the derivative is 0 while executing the Newton-Raphson method
  /// @throws Exception if the Newton-Raphson method fails to converge in the
  double? calculate() {
    final double years = daysBetween(details.start!, details.end!) / daysInYear;
    if (details.maxAmount == 0) {
      return -1; // Total loss
    }
    guess = guess ?? (details.total! / details.deposits!) / years;
    return NewtonRaphson.withFunctions(presentValue, derivative)
        .findRoot(guess!);
  }
}

extension Ex on double {
  double toPrecision(int n) => double.parse(toStringAsFixed(n));
}
